### PR TITLE
Decode tx

### PR DESCRIPTION
**Pull request by 7sne**

### CODE REVIEW

File: src/lib/decodeEvent.ts [hunk 1]
```diff
@@ -44,6 +45,8 @@ export function decodeEvent(
   iface: Interface,
   eventProps: EventProps,
 ): DecodedEventResult {
+  if (!isHex(eventProps.data)) eventProps.data = '0x' + eventProps.data;
```
Comment (LogvinovLeon) [inline src/lib/decodeEvent.ts:48]: This is kinda user input validation/adjustment. I think we should handle it earlier. As soon as we get it. So in the component. The data passed to logic functions should be already correct and the input from the user should be fixed as fast as possible.

File: src/lib/decodeTx.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import { TransactionFactory, TypedTransaction } from '@ethereumjs/tx';
+import { toBuffer } from 'ethereumjs-util';
+
+export interface DecodedTx {
+  tx: TypedTransaction;
+  senderAddr: string;
+}
+
+export function decodeTx(rawTx: string): DecodedTx {
+  let prefixed = rawTx.trim();
```
Comment (LogvinovLeon) [inline src/lib/decodeTx.ts:10]: as above

File: src/lib/decodeTx.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+  if (rawTx.trim().substring(0, 2) !== '0x') {
```
Comment (LogvinovLeon) [inline src/lib/decodeTx.ts:12]: aa

File: src/lib/decodeTx.test.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      '0x02f87301398459682f0085071c42979e82520894c9e7a3937cf24ed932b69c4545a20a5966a73fb58803311fc80a57000080c001a005a5ee2b0994acdd28f43b8fc2594cd702ada33f1f319308352e401e9b493ef9a011dd2830018dbf05a7f37860d6cff9f530b073836985005083469a86d805a7a2',
+    );
+
+    expect(actual.tx).toBeA(FeeMarketEIP1559Transaction);
+    expect(actual.senderAddr).toEqual(
+      '0xa784398ad2a43b8ba13b0eb64664778f088bf6d2',
+    );
+  });
+
+  it('handles EIP2930 transaction', () => {
```
Comment (LogvinovLeon) [inline src/lib/decodeTx.test.ts:33]: We have two types of tests that check full decoding and one that does it differently.
If there is a bug in decoding library, but it is consistent with encoding - we will not see it in the last test. What I'm proposing is asserting types & sender addresses in all three tests and also adding snapshots after. That way any time they change the format on any of the types - we will notice

**Review by LogvinovLeon — CHANGES_REQUESTED** [review]

**Comment by LogvinovLeon** [comment]
Also - lets go a little further in our decoding efforts.
Screen Shot 2022-03-31 at 11 15 15 AM

There are some fields for which we know the format for sure and we don't need to display them as obscure hashes.
`nonce`, `gasPrice`, `gasLimit` should be displayed as ints.
`value` should be also int, but divided by `10^18`.

Yes, it will require some more FE layout, but it will also make our FE test nicer as we will be able to assert specific fields with human readable names instead of json blob with hexes.

**Comment by LogvinovLeon** [comment]
Let's do the above comment in the subsequent PR as if would contain significant part of the scope of another issue: https://linear.app/deth/issue/DET-105/tooltip-to-choose-value-display-format

File: src/lib/decodeTx.test.ts [hunk 5]
```diff
@@ -14,6 +14,7 @@ describe(decodeTx.name, () => {
     );
 
     expect(actual.tx).toBeA(Transaction);
+    expect(actual.tx).toMatchSnapshot();
     expect(actual.senderAddr).toEqual(
       '0x67835910d32600471f388a137bbff3eb07993c04',
     );
```
Comment (LogvinovLeon) [inline src/lib/decodeTx.test.ts:20]:
```suggestion
    expect(actual.tx).toBeA(Transaction);
    expect(actual.senderAddr).toEqual(
      '0x67835910d32600471f388a137bbff3eb07993c04',
    );
    expect(actual.tx).toMatchSnapshot();
```

Let's have a convention to have snapshot assertions as the last assertion of the group.
They usually produce less readable output and serve as a last resort/catch all assertion.

File: src/lib/decodeTx.test.ts [hunk 6]
```diff
@@ -25,6 +26,7 @@ describe(decodeTx.name, () => {
...
 
     expect(actual.tx).toBeA(FeeMarketEIP1559Transaction);
+    expect(actual.tx).toMatchSnapshot();
     expect(actual.senderAddr).toEqual(
       '0xa784398ad2a43b8ba13b0eb64664778f088bf6d2',
     );
```
Comment (LogvinovLeon) [inline src/lib/decodeTx.test.ts:32]:
```suggestion
    expect(actual.tx).toBeA(FeeMarketEIP1559Transaction);
    expect(actual.senderAddr).toEqual(
      '0xa784398ad2a43b8ba13b0eb64664778f088bf6d2',
    );
    expect(actual.tx).toMatchSnapshot();
```

File: src/lib/prefixHex.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+import { isHex } from './decodeHex';
+
+export function prefixHexIf0xMissing(data: string): string {
+  if (!isHex(data)) return '0x' + data;
+  else return data;
+}
```
Comment (LogvinovLeon) [inline src/lib/prefixHex.ts:6]: Just use `addHexPrefix` in `ethereumjs-util`

Comment (LogvinovLeon) [reply src/lib/prefixHex.ts:6]: https://github.com/ethereumjs/ethereumjs-util/blob/master/docs/modules/_bytes_.md#const-addhexprefix

Comment (LogvinovLeon) [reply src/lib/prefixHex.ts:6]: You also have `stripHexPrefix` and `padToEven` there

File: src/lib/toEvenHex.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    } else {
+      return '0x' + str;
+    }
+  } else {
+    if (str.length % 2 !== 0) {
+      return '0' + str;
+    }
+  }
+  return str;
+}
```
Comment (LogvinovLeon) [inline src/lib/toEvenHex.ts:20]: Remove and use `ethereumjs-util` helpers

File: src/lib/toEvenHex.test.ts [hunk 9]
```diff
@@ -0,0 +1,13 @@
+import { expect } from 'earljs';
+
+import { toEvenHex } from './toEvenHex';
+
+describe(toEvenHex.name, () => {
+  it('converts odd hex to even hex', () => {
+    expect(toEvenHex('0x0')).toEqual('0x0');
```
Comment (LogvinovLeon) [inline src/lib/toEvenHex.test.ts:7]:
```suggestion
    expect(toEvenHex('0x0')).toEqual('0x00');
```

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -25,37 +25,15 @@ jobs:
...
+            hashFiles('**/yarn.lock') }}
           restore-keys: ${{ runner.OS }}-${{ matrix.node }}-yarn-cache-
 
       - run: yarn --no-progress --non-interactive --frozen-lockfile
 
-      - run: yarn build
-      - run: yarn lint
+      - name: 'TypeCheck, Lint and Build'
+        # NextJS typechecks and lints before the build
+        run: yarn build
```
Comment (LogvinovLeon) [inline .github/workflows/ci.yml:37]:
```suggestion
      - run: yarn build
```

**Review by LogvinovLeon — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 10 above)
Comment (hasparus) [reply .github/workflows/ci.yml:37]: Will this work @LogvinovLeon 
? The object starts with "name"

**Pull request merged**